encoding/Expression.java

```diff
@@ -0,0 +1,200 @@
+
+        public static final Operation PLUS_DOUBLE_LONG = new Operation("+", Encoding.ValueType.DOUBLE, Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG);
+        public static final Operation MINUS_DOUBLE_LONG = new Operation("-", Encoding.ValueType.DOUBLE, Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG);
+        public static final Operation TIMES_DOUBLE_LONG = new Operation("*", Encoding.ValueType.DOUBLE, Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG);
+        public static final Operation DIV_DOUBLE_LONG = new Operation("/", Encoding.ValueType.DOUBLE, Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG);
+        public static final Operation POW_DOUBLE_LONG = new Operation("^", Encoding.ValueType.DOUBLE, Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG);
+
+        public static final Operation PLUS_LONG_DOUBLE = new Operation("+", Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE);
+        public static final Operation MINUS_LONG_DOUBLE = new Operation("-", Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE);
+        public static final Operation TIMES_LONG_DOUBLE = new Operation("*", Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE);
+        public static final Operation DIV_LONG_DOUBLE = new Operation("/", Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE);
+        public static final Operation POW_LONG_DOUBLE = new Operation("^", Encoding.ValueType.DOUBLE, Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE);
+
+        private final String symbol;
+        private final Encoding.ValueType<?> arg1Type;
+        private final Encoding.ValueType<?> arg2Type;
+
+        Operation(String symbol, Encoding.ValueType<?> returnType, Encoding.ValueType<?> arg1Type, Encoding.ValueType<?> arg2Type) {
+            super(returnType);
+            this.symbol = symbol;
+            this.arg1Type = arg1Type;
+            this.arg2Type = arg2Type;
+            Expression.register(this);
+        }
+    }
+
+    public static class Function extends Expression {
+        public static final Function FLOOR_DOUBLE = new Function("floor", Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE);
+        public static final Function CEIL_DOUBLE = new Function("ceil", Encoding.ValueType.LONG, Encoding.ValueType.DOUBLE);
+
+        private final String symbol;
+        private final List<Encoding.ValueType<?>> argumentTypes;
+
+        Function(String symbol, Encoding.ValueType<?> returnType, Encoding.ValueType<?>... argumentTypes) {
+            super(returnType);
+            this.symbol = symbol;
+            this.argumentTypes = list(argumentTypes);
+            Expression.register(this);
+        }
+
+        public List<Encoding.ValueType<?>> argumentTypes() {
+            return argumentTypes;
+        }
+    }
+
+    public static class ThingVar extends Expression {
```

Comment:
I can do away with having a separate `ThingVar` and `ValueVar` if preferred - they make little difference at the traversal level.

Reply:
Just plain var should be fine. 
Now that the ValueHolder interface exists, it should be possible.